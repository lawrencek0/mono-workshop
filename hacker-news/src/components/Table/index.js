import React from 'react';
import Button from '../Button';
import './index.css';

const Table = ({list, onDismiss}) => {
  return (
    <div className="table">
      {list.map((item) => (
        <div key={item.objectID} className="table-row">
          <span className="largeColumn">
            <a href={item.url}>{item.title}</a>
          </span>
          <span className="midColumn">{item.author}</span>
          <span className="smallColumn">{item.num_comments}</span>
          <span className="smallColumn">{item.points}</span>
          <span className="smallColumn">
            <Button
              onClick={() => onDismiss(item.objectID)}
              className="button-inline"
            >
              Dismiss
            </Button>
          </span>
        </div>
      ))}
    </div>
  )
};

export default Table;