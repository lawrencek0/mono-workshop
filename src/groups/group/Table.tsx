import React from 'react';
import { useTable, usePagination } from 'react-table';
import { GroupUserResource } from 'resources/GroupResource';
import styled from 'styled-components/macro';
import tw from 'tailwind.macro';
import { FlatButton, PrimaryButton } from 'shared/buttons';
import { useFetcher } from 'rest-hooks';

export const Table: React.FC<{
    users: GroupUserResource[];
    canModify: boolean;
    groupId?: string;
}> = ({ users, groupId, canModify }) => {
    const {
        getTableProps,
        getTableBodyProps,
        headerGroups,
        prepareRow,
        page,
        canPreviousPage,
        canNextPage,
        gotoPage,
        nextPage,
        previousPage,
        pageCount,
    } = useTable(
        {
            columns: [
                { Header: 'Name', id: 'name', accessor: r => `${r.firstName} ${r.lastName}` },
                { Header: 'Role', id: 'role', accessor: r => r.role as string },
            ],
            data: users,
        },
        usePagination,
    );
    const update = useFetcher(GroupUserResource.partialUpdateShape());
    const deleter = useFetcher(GroupUserResource.deleteShape());

    const handleRoleChange = (user: GroupUserResource) => async (
        e: React.ChangeEvent<HTMLSelectElement>,
    ): Promise<void> => {
        await update({ id: user.id, groupId }, { role: e.target.value as 'member' | 'mod' | 'owner' });
    };

    const handleUserDelete = async (user: GroupUserResource): Promise<void> => {
        await deleter({ id: user.id, groupId }, undefined);
    };

    return (
        <div>
            <StyledTable {...getTableProps()}>
                <thead>
                    {headerGroups.map(headerGroup => (
                        <tr key={undefined} {...headerGroup.getHeaderGroupProps()}>
                            {headerGroup.headers.map(column => (
                                <StyledTh key={undefined} {...column.getHeaderProps()}>
                                    {column.render('Header')}
                                </StyledTh>
                            ))}
                        </tr>
                    ))}
                </thead>
                <tbody {...getTableBodyProps()}>
                    {page.map(row => {
                        prepareRow(row);
                        return (
                            <tr key={undefined} {...row.getRowProps()}>
                                {row.cells.map(cell => {
                                    return (
                                        <StyledTd key={undefined} {...cell.getCellProps()}>
                                            {cell.render('Cell')}
                                        </StyledTd>
                                    );
                                })}
                                {canModify && (
                                    <>
                                        <StyledTd>
                                            <select
                                                defaultValue={row.original.role}
                                                onChange={handleRoleChange(row.original as GroupUserResource)}
                                            >
                                                <option value="member">Member</option>
                                                <option value="mod">Mod</option>
                                                <option value="owner">Admin</option>
                                            </select>
                                        </StyledTd>
                                        <StyledTd>
                                            <PrimaryButton
                                                variant="danger"
                                                onClick={() => handleUserDelete(row.original as GroupUserResource)}
                                            >
                                                Delete?
                                            </PrimaryButton>
                                        </StyledTd>
                                    </>
                                )}
                            </tr>
                        );
                    })}
                </tbody>
            </StyledTable>
            <Pagination>
                <FlatButton onClick={() => gotoPage(0)} disabled={!canPreviousPage}>
                    {'<<'}
                </FlatButton>
                <FlatButton onClick={() => previousPage()} disabled={!canPreviousPage}>
                    {'<'}
                </FlatButton>
                <FlatButton onClick={() => nextPage()} disabled={!canNextPage}>
                    {'>'}
                </FlatButton>
                <FlatButton onClick={() => gotoPage(pageCount - 1)} disabled={!canNextPage}>
                    {'>>'}
                </FlatButton>
            </Pagination>
        </div>
    );
};

const Pagination = styled.div`
    ${tw`flex justify-between py-4`}
`;

const StyledTable = styled.table`
    ${tw`text-left w-full border-collapse`}
`;

const StyledTh = styled.th`
    ${tw`py-4 px-6 font-bold uppercase text-sm text-gray-dark border-b`}
`;

export const StyledTd = styled.td`
    ${tw`py-4 px-6 border-b border-gray-500`}
`;
