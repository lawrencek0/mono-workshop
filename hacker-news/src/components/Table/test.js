import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Table from '../Table';

describe('Table', () => {

  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
      { title: '2', author: '2', num_comments: 2, points: 2, objectID: 'z' },
    ],
    sortKey: 'TITLE',
    isSortReverse: false,
  };

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Table {...props} />, div);
  });

  test('snapshots', () => {
    const component = renderer.create(
      <Table {...props} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows two items in the list', () => {
    const element = shallow(
      <Table {...props} />
    );

    expect(element.find('.table-row').length).toBe(2);
  });

});

