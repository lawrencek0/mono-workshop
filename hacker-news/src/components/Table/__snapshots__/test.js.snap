exports[`Table snapshots 1`] = `
<div
  className="table">
  <div
    className="table-row">
    <span
      className="largeColumn">
      <a
        href={undefined}>
        1
      </a>
    </span>
    <span
      className="midColumn">
      1
    </span>
    <span
      className="smallColumn">
      1
    </span>
    <span
      className="smallColumn">
      2
    </span>
    <span
      className="smallColumn">
      <button
        className="button-inline"
        onClick={[Function]}
        type="button">
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row">
    <span
      className="largeColumn">
      <a
        href={undefined}>
        2
      </a>
    </span>
    <span
      className="midColumn">
      2
    </span>
    <span
      className="smallColumn">
      2
    </span>
    <span
      className="smallColumn">
      2
    </span>
    <span
      className="smallColumn">
      <button
        className="button-inline"
        onClick={[Function]}
        type="button">
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
